redesign and make 3.1 ready (#7)
ibm-cloud-architecture/terraform-icp-ibmcloud: templates/icp-ee/README.md

# Terraform ICP IBM Cloud

This Terraform example configurations uses the [IBM Cloud  provider](https://ibm-cloud.github.io/tf-ibm-docs/index.html) to provision virtual machines on IBM Cloud Infrastructure (SoftLayer)
and [Terraform Module ICP Deploy](https://github.com/ibm-cloud-architecture/terraform-module-icp-deploy) to prepare VSIs and deploy [IBM Cloud Private](https://www.ibm.com/cloud-computing/products/ibm-cloud-private/) version 3.1.0 or later in a Highly Available configuration.  This Terraform template automates best practices learned from installing ICP on IBM Cloud Infrastructure.

## Deployment overview
This template creates an environment where

>  - Cluster is deployed on private network and is accessed loadbalancers.

Reviewer: ```suggestion
 - Cluster is deployed on private network and is accessed through load balancers.
```

 - Dedicated management node
 - Dedicated boot node
 - SSH access from public network is enabled on boot node only
 - Optimised VM sizes
 - IBM File Storage providing shared storage for master nodes

> 

Reviewer: No VA node

## Architecture Diagram

![Architecture](../../static/icp_ibmcloud.png)

## Pre-requisites

* Working copy of [Terraform](https://www.terraform.io/intro/getting-started/install.html)
  * As of this writing, IBM Cloud Terraform provider is not in the main Terraform repository and must be installed manually.  See [these steps](https://ibm-cloud.github.io/tf-ibm-docs/index.html#using-terraform-with-the-ibm-cloud-provider).  The templates have been tested with Terraform version 0.11.7 and the IBM Cloud provider version 0.11.3.
* The template is tested on VSIs based on Ubuntu 16.04.  RHEL is not supported in this automation.

### Environment preparation

There are two options to provide the IBM Cloud Private binaries necessary to install:
1. Download the IBM Cloud Private docker and installation binaries and save them to the `icp-install` directory.

> 2. (Preferred) Create an NFS mount point with IBM Cloud File Storage and upload the binaries there. See below to the [Setup IBM Cloud File Storage](#setup-ibm-cloud-file-storage) for this purpose.

Reviewer: ```suggestion
2. (Preferred) Create an NFS mount point with IBM Cloud File Storage and upload the binaries there. See below to the [Setup IBM Cloud File Storage](#setup-ibm-cloud-file-storage-to-host-icp-binaries) for this purpose.
```

Author: Top marks for spotting that one!

3. Create an HTTP endpoint and upload the binaries there.

### Using the Terraform templates

1. git clone the repository

1. Navigate to the template directory `templates/icp-ee`

1. Create a `terraform.tfvars` file to reflect your environment.  Please see [variables.tf](variables.tf) and below tables for variable names and descriptions.  Here is an example `terraform.tfvars` file:

```
sl_username = "<my username>"
sl_api_key = "<my api key>"
key_name = ["my-ssh-key"]
datacenter = "dal13"
os_reference_code = "UBUNTU_16_64"
icp_inception_image = "ibmcom/icp-inception-amd64:3.1.0-ee"
image_location = "nfs:fsf-dal1301i-fz.adn.networklayer.com:/IBMnnSVnnnn_n/data01/files/icp/ibm-cloud-private-x86_64-3.1.0.tar.gz"
```

1. Run `terraform init` to download depenencies (modules and plugins)

1. Run `terraform plan` to investigate deployment plan

1. Run `terraform apply` to start deployment.

### Automation Notes

#### What does the automation do
1. Create security groups and rules for cluster communication as declared in [security_group.tf](security_group.tf)

> 1. Create Loadbalancers for Proxy and Control plane

Reviewer: ```suggestion
1. Create load balancers for Proxy and Control plane
```

2. Create IBM File Storage for master nodes shared storage
1. Create the virtual machines as defined in `variables.tf` and `terraform.tfvars`

>   - Use cloud-init to add a user `icpdeploy` with a randomly generated ssh-key

Reviewer: ```suggestion
   - Use cloud-init to add a user `icpdeploy` with a randomly generated ssh-key
```

>   - Configure a separate hard disk to be used by docker

Reviewer: ```suggestion
   - Configure a separate hard disk to be used by docker
```

>   - Configure the shared storage on master nodes

Reviewer: ```suggestion
   - Configure the shared storage on master nodes
```

1. Handover to the [icp-deploy](https://github.com/ibm-cloud-architecture/terraform-module-icp-deploy) terraform module as declared in the [icp-deploy.tf](icp-deploy.tf) file

#### What does the icp deploy module do
1. It uses the provided ssh key which has been generated for the `icpdeploy` user to ssh from the terraform controller to all cluster nodes to install ICP prerequisites
2. It generates a new ssh keypair for ICP Boot(master) node to ICP cluster communication and distributes the public key to the cluster nodes. This key is used by the ICP Ansible installer.

> 3. It populates the necessary /etc/hosts file on the boot node

Reviewer: ```suggestion
3. It populates the necessary `/etc/hosts` file on the boot node
```

4. It generates the ICP cluster hosts file based on information provided in [icp-deploy.tf](icp-deploy.tf)

> 5. 4. It generates the ICP cluster `config.yaml` file based on information provided in [icp-deploy.tf](icp-deploy.tf)

Reviewer: ```suggestion
5. It generates the ICP cluster `config.yaml` file based on information provided in [icp-deploy.tf](icp-deploy.tf)
```

| `private_vlan_router_hostname` | no | Private VLAN router to place all VSIs behind.  e.g. bcr01a. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. This option should be used when setting `private_network_only` to true along with `private_vlan_number` using a private VLAN that is routed with a Gateway Appliance. |
| `private_vlan_number` | no | Private VLAN number to place all VSIs on.  e.g. 1211. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. This option should be used when setting `private_network_only` to true along with `private_vlan_router_hostname`, using a private VLAN that is routed with a Gateway Appliance.|
| `public_vlan_router_hostname` | no | Public VLAN router to place all VSIs behind.  e.g. fcr01a. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. |
| `public_vlan_number` | no | Public VLAN number to place all VSIs on.  e.g. 1211. See Network > IP Management > VLANs in the portal. Leave blank to let the system choose. |
| `icppassword` | no | ICP administrator password.  One will be generated if not set. |

> 

Reviewer: ```suggestion
| `deployment` | no | Identifier prefix added to the host names of all your infrastructure resources for organising/naming ease |
```
